source/dhtslib/threadpool.d

```diff
@@ -0,0 +1,482 @@
+/**
+This module provides multithreading support through htslib. It allows for mapping delagates,
+and functions via a range interface with `parallelMap`. Crucially, the htslib task pool performs
```

Comment:
So my intended use is primarily to be able to do work in parallel in an in-order fashion. This can be done via `std.parallelism` though I have never been able to get it to work for me. Though this also can provide the advantage of sharing the thread pool with readers and writers. Then whatever jobs are taking the most work will be taking a majority of the threads. So if i have 32 cores to deal with, if my work to be done is more compute intensive than the need to decompress and do file I/O, it can use a majority of those threads.

So my intended use case would be something like the following:
```d
// set up global pool
enforceGlobalPoolThreads(32);

// use global pool for multi-threaded decompression
auto vcfr = VcfReader("a.vcf.gz");
vcfr.setThreadPool(globalPool);

// use global pool for multi-threaded compression/writing
auto vcfw = VcfWriter("b.vcf.gz", vcfr.header);
vcfw.setThreadPool(globalPool);
vcfw.writeHeader();

// perform work in parallel but in order
auto range = vcfr.parallelMap((rec) {
    // Even the reading of the gff3 could use the same thread pool
    rec.filterAnnnotationsToPrincipalIsoform(rec.infos["ANN"], "gencode.gff3.gz");
    return rec;
})

// Though this loop is single threaded, the range it is pulling from is not
// And the writer is writing the results in parallel
// Since data is returned in order, no need to re-sort later
foreach(rec;range) {
    vcfw.write(rec);
}
```

In a similar vein, I could use this in `fade` for preventing the need to resort bam files after artifact identification before filtering artifact reads, while retaining the ability to do work in parallel.

Reply:
Makes sense. When you say have never been able to get std parallelism to work for you, what does that mean with respect to the fact that you have imported its code? Do you expect it to work? S is it tested already and confirmed working?

Reply:
I have never been able to get `std.parallelism` 's provided functions to work for my purposes. `std.parallelism`'s `TaskPool.map` does in-order but evaluates lazily, and in my testing didn't work very well. `std.parallelism`'s `TaskPool.amap` is similar but an eager algorithm but requires a `RandomAccessRange` as input. `std.parallelism`'s `parallel foreach` function works pretty well but doesn't output a range and is not in-order. 

The code I have ported from `std.parallelism` works as intended. It eagerly evaluates tasks in-order and is more leinient to which delegates it can run. Though it is likely not quite as safe, due to some more manual memory management. 

Though the user shouldn't have to worry about it. Notably, the code I wrote is actually executed serially,  only the tasks are executed in parallel via htslib and htslib's `hts_tpool_next_result_wait` makes it easy to synchronously grab the next ready result as it becomes available. Making wrapping it in a range pretty easy.

Reply:
Additionally, my `parallelMap` function can use a simple `InputRange` as input.